proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+
+For example the custom `FileLoadError` can wrap the `std::io::Error`.
+This way the file loading function doesn't have to know about all the possible file system failures
+described by the `std::io::Error`.
+When it gets one, it simply wraps it in the `FileLoadError` and returns it.
+If a higher level function is interested, what caused the `FileLoadError`, it can call its `source`
+method and get its cause.
+
+`source` should return `None` when the error is the root cause and doesn't wrap another error.
+It's also impossible to return a wrapped error when it doesn't implement `Error`.
+
+### The bevavior of the `Display`
+
+The `Display` implementation should briefly describe the error.
+It should never include information provided by the error returned from the `source`.
+Its output should be formatted as a single line, preferably a single sentence.
+If it's a single sentence, it should start with a lowercase letter and end with no interpunction.
+
+Examples:
+
+`failed to load the file '/tmp/file'` **good**
+
+`failed to load the file '/tmp/file': file not found`
+**bad**: includes information provided by its `source`
+
+`Failed to load the file '/tmp/file'!`
+**bad**: a single sentence starting with a capital letter and ending with interpunction
+
+`Failed to load the file. Path: '/tmp/file'.`
+**ugly**: it has two sentences
+
+## Printing a chain of `Error`s
+
+When all the errors in a chain behave correctly, insightful and well formatted messages can be
+easily provided to the user.
+The error can be iterated by recursively calling `source` to print each layer and provide the full
+chain of errors that caused the failure.
+Due to strict `Display` implementation rules there's a lot of creative freedom when it comes to
+chain formatting.
+The `source` traverasal can be done manually, with `anyhow::Chain` or with very similar
+`std::error::Chain` as soon as it's stabilized.
+
+An example of the manual chain printing:
+
+```rust
+println!("An error occured: {}", error);
+let mut parent = &error;
+while let Some(cause) = parent.source() {
+    println!("  caused by: {}", cause);
+    parent = cause;
+}
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file
+  caused by: failed to load the file '~/.app_config'
+  caused by: file not found
+```
+
+An example of the chain printing with `anyhow::Chain`:
+
+```rust
+print!("An error occured: {}", error);
+anyhow::Chain::new(&error).for_each(|e| print!(": {}", e);
+println!();
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file: failed to load the file '~/.app_config': file not found
+```
+
+# Recommendation
+
+These examples of building errors serve as a practical guide.
+There are some common patterns of error structures and all of them can be implemented conveniently
+and correctly using `thiserror` and `anyhow`.
+
+All the examples use `enum`s, but it's fine to use `struct`s instead when there's only one variant.
```

Comment:
The recommendation of `enum` and `struct` is already outlined in the previous error handling proposal. So I would just reference that and not elaborate here.

Reply:
These are just two sentences, do you think that it's worth referencing the other document? When it's here, it resolves any confusion in ~4 seconds of reading.

Reply:
Fair point. I'm realising that it's actually the line below that alerted me: " Usually it's not very useful, because it reduces flexibility, but doesn't simplify code much." This relates to a recommendation that was already made in the previous proposal and I don't think adds anything here.

Do you agree?

Reply:
Removed